drivers/power_gating/power_gating_intel.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <drivers/gpio.h>
+#include <pm/device.h>
+#include <pm/device_runtime.h>
+#include <pm/power_gating.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(power_mgmt_ACE, LOG_LEVEL_INF);
+
+static int poll_for_register_delay(uint32_t reg, uint32_t mask, uint32_t val, uint64_t us)
+{
+	if (!WAIT_FOR((sys_read32(reg) & mask) != val, us, k_busy_wait(1))) {
+		LOG_ERR("%s poll timeout reg %u mask %u val %u us %u", __func__, reg, mask, val,
+			(uint32_t)us);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+int device_set_power_enable(struct pm_pg_registers reg, bool power_enable, uint64_t us)
```

Comment:
```suggestion
int device_set_power_enable(struct pm_pg_registers reg, bool power_enable, uint64_t us)
```
I think this API is problematic, it is not integrated in the device model. I mean, if one calls this API directly, Zephyr won't be able to track the device power state, devices under the domain are not notified that they are powered and also can lead to problems with device runtime.

Reply:
This is just to avoid code duplication. Every driver has to implement the same mechanism as a part of power gating cycle, but it also has to do some driver specific stuff. It's not intended to be used separately. Current implementation also allows the driver to be turned off without notifying zephyr power management system. Maybe a comment notice there would suffice?
It's similar to current sof implementation, where there is a common function (platform_pm_runtime_put/get), though this approach is limited to power gating and instead of being a big platform dependant switch we can use devicetree to make it a bit more universal.

Reply:
Gotcha. Thought SoF API is similar to Zephyr's API, the way it works is different. On SoF you have this pm runtime driver that centralize the power management for all devices and each devices are identified by handles. On Zephyr, each device controls its own power management and they are identified by the pointer to them. e.g: `pm_device_runtime_get(dev)`.

It is ok (and actually recommended) devices suspend/resume by themselves but they have to do it using pm device runtime  APIs, so Zephyr can do the necessary bookkeeping.